contracts/upgradeable_contracts/BasicAMBMediator.sol

```diff
@@ -0,0 +1,134 @@
+    * @dev Sets the mediator contract address from the other network. Only the owner can call this method.
+    * @param _mediatorContract the address of the mediator contract.
+    */
+    function setMediatorContractOnOtherSide(address _mediatorContract) external onlyOwner {
+        _setMediatorContractOnOtherSide(_mediatorContract);
+    }
+
+    /**
+    * @dev Sets the gas limit to be used in the message execution by the AMB bridge on the other network.
+    * This value can't exceed the parameter maxGasPerTx defined on the AMB bridge.
+    * Only the owner can call this method.
+    * @param _requestGasLimit the gas limit for the message execution.
+    */
+    function setRequestGasLimit(uint256 _requestGasLimit) external onlyOwner {
+        _setRequestGasLimit(_requestGasLimit);
+    }
+
+    /**
+    * @dev Get the AMB interface for the bridge contract address
+    * @return AMB interface for the bridge contract address
+    */
+    function bridgeContract() public view returns (IAMB) {
+        return IAMB(addressStorage[BRIDGE_CONTRACT]);
+    }
+
+    /**
+    * @dev Tells the mediator contract address from the other network.
+    * @return the address of the mediator contract.
+    */
+    function mediatorContractOnOtherSide() public view returns (address) {
+        return addressStorage[MEDIATOR_CONTRACT];
+    }
+
+    /**
+    * @dev Tells the gas limit to be used in the message execution by the AMB bridge on the other network.
+    * @return the gas limit for the message execution.
+    */
+    function requestGasLimit() public view returns (uint256) {
+        return uintStorage[REQUEST_GAS_LIMIT];
+    }
+
+    /**
+    * @dev Stores a valid AMB bridge contract address.
+    * @param _bridgeContract the address of the bridge contract.
+    */
+    function _setBridgeContract(address _bridgeContract) internal {
+        require(AddressUtils.isContract(_bridgeContract));
+        addressStorage[BRIDGE_CONTRACT] = _bridgeContract;
+    }
+
+    /**
+    * @dev Stores the mediator contract address from the other network.
+    * @param _mediatorContract the address of the mediator contract.
+    */
+    function _setMediatorContractOnOtherSide(address _mediatorContract) internal {
+        addressStorage[MEDIATOR_CONTRACT] = _mediatorContract;
+    }
+
+    /**
+    * @dev Stores the gas limit to be used in the message execution by the AMB bridge on the other network.
+    * @param _requestGasLimit the gas limit for the message execution.
+    */
+    function _setRequestGasLimit(uint256 _requestGasLimit) internal {
+        require(_requestGasLimit <= maxGasPerTx());
+        uintStorage[REQUEST_GAS_LIMIT] = _requestGasLimit;
+    }
+
+    /**
+    * @dev Sets the unique identifier to be used for the next message to be sent to the AMB bridge.
+    * @param _hash representing the next message unique identifier.
+    */
+    function setNonce(bytes32 _hash) internal {
+        bytesStorage[NONCE] = abi.encodePacked(_hash);
```

Comment:
Here, and in all similar places (like `setTransactionHash()`, `setFailedMessageDataHash(bytes32)`), what is the purpose of storing data in `bytes` rather than in `bytes32`?
Seems to me, that if the value is always `32` bytes long, storing it as a `bytes32` will save some gas.

Reply:
You are right, storing as `bytes32` will save some gas. The issue is that in the `EternalStorage` there is no mapping for storing `mapping(bytes32 => bytes32)` that's why we use the workaround to store it as `bytes` in order to not modify the storage layout.

Reply:
We can modify `EternalStorage` by adding a `mapping(bytes32 => bytes32)` at the end of the file? This shouldn't change the storage layout.

Reply:
Ok, this is not possible if the plan is to upgrade current bridge contracts to the new mediator contracts.

Reply:
By the way, can't we use `uintStorage` for storing `bytes32` values instead?

Reply:
Thanks! Storing it in `uintStorage` saves some gas. I updated how the nonce is stored. Regarding `setTransactionHash()` and `setFailedMessageDataHash`, as they are part of AMB bridge itself, I think it makes sense to create an issue for this improvement to properly evaluate and apply the changes separately.  https://github.com/poanetwork/tokenbridge-contracts/pull/367/commits/4a615de4b340218245f154faeb5c22c7b14195bf

Reply:
Great. Created #373